src/history/hs_rec.c

```diff
@@ -464,54 +432,66 @@ __wt_hs_insert_updates(
+         * history store.
          */
-        if (oldest_upd->type == WT_UPDATE_TOMBSTONE && oldest_upd == first_non_ts_upd &&
-          !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS)) {
-            /* We can only delete history store entries that have timestamps. */
-            WT_ERR(__wt_hs_delete_key_from_ts(session, hs_cursor, btree->id, key, 1, true));
-            WT_STAT_CONN_INCR(session, cache_hs_key_truncate_non_ts);
-            WT_STAT_DATA_INCR(session, cache_hs_key_truncate_non_ts);
-            F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
-        } else if (first_non_ts_upd != NULL && !F_ISSET(first_non_ts_upd, WT_UPDATE_CLEARED_HS) &&
-          (list->ins == NULL || ts_updates_in_hs)) {
-            WT_ERR(__wt_hs_delete_key_from_ts(session, hs_cursor, btree->id, key, 1, true));
-            WT_STAT_CONN_INCR(session, cache_hs_key_truncate_non_ts);
-            WT_STAT_DATA_INCR(session, cache_hs_key_truncate_non_ts);
-            F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
+        if (oldest_upd->type == WT_UPDATE_TOMBSTONE) {
+            if (out_of_order_ts_upd != NULL && out_of_order_ts_upd->start_ts < oldest_upd->start_ts)
+                fix_ts_upd = out_of_order_ts_upd;
+            else
+                fix_ts_upd = oldest_upd;
+
+            if (!F_ISSET(fix_ts_upd, WT_UPDATE_FIXED_HS)) {
+                /* Delete and reinsert any update of the key with a higher timestamp.
+                 */
+                WT_ERR(__wt_hs_delete_key_from_ts(
+                  session, hs_cursor, btree->id, key, fix_ts_upd->start_ts + 1, true));
+                F_SET(fix_ts_upd, WT_UPDATE_FIXED_HS);
+            }
         }
 
-        WT_ERR(__hs_next_upd_full_value(session, &modifies, NULL, full_value, &upd));
+        WT_ERR(__hs_next_upd_full_value(session, &updates, NULL, full_value, &upd));
 
         hs_inserted = squashed = false;
 
         /*
          * Flush the updates on stack. Stopping once we run out or we reach the onpage upd start
-         * time point, we can squash modifies with the same start time point as the onpage upd away.
+         * time point, we can squash updates with the same start time point as the onpage update
+         * away.
          */
-        for (; modifies.size > 0 &&
+        for (; updates.size > 0 &&
              !(upd->txnid == list->onpage_upd->txnid &&
                upd->start_ts == list->onpage_upd->start_ts);
              tmp = full_value, full_value = prev_full_value, prev_full_value = tmp,
              upd = prev_upd) {
             WT_ASSERT(session, upd->type == WT_UPDATE_STANDARD || upd->type == WT_UPDATE_MODIFY);
 
-            tw.durable_start_ts = upd->durable_ts;
-            tw.start_ts = upd->start_ts;
-            tw.start_txn = upd->txnid;
             tombstone = NULL;
-            __wt_modify_vector_peek(&modifies, &prev_upd);
+            __wt_update_vector_peek(&updates, &prev_upd);
+
+            if (out_of_order_ts_updates.size > 0) {
```

Comment:
Can we make it a macro?

Reply:
Why do you think a macro is better?

Reply:
Just for the sake of abstraction - if we somehow change the variable names in underlying implementation of vector, we don't have to change the code. But then again it's just a styling thing.